feat: Add MySQL instance address attribute

arielvalentin opened the pull request:
Fixes # N/A

## Changes

<!--
Please provide a brief description of the changes here.


Note: if the PR is touching an area that is not listed in the [existing areas](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/README.md), or the area does not have sufficient [domain experts coverage](https://github.com/open-telemetry/semantic-conventions/blob/main/.github/CODEOWNERS), the PR might be tagged as [experts needed](https://github.com/open-telemetry/semantic-conventions/labels/experts%20needed) and move slowly until experts are identified.
-->

At GitHub we run MySQL in multi-node clusters and are often interested in knowing the hostname of a node that is executing a query or command.

In order to acheive this we added an attribute to the Trilogy instrumentation named `db.mysql.instance.address` which is populated with the hostname of the node:

<https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/487>

This change adds the attribute to the database model as a MySQL specific connection attribute.

## Merge requirement checklist

* [x] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [ ] [CHANGELOG.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CHANGELOG.md) updated for non-trivial changes.
* [ ] [schema-next.yaml](https://github.com/open-telemetry/semantic-conventions/blob/main/schema-next.yaml) updated with changes to existing conventions.

pyohannes commented on docs/database/mysql.md:
```diff
@@ -0,0 +1,24 @@
+| Attribute  | Type | Description  | Examples  | Requirement Level |
+|---|---|---|---|---|
+| [`db.jdbc.driver_classname`](database-spans.md) | string | The fully-qualified class name of the [Java Database Connectivity (JDBC)](https://docs.oracle.com/javase/8/docs/technotes/guides/jdbc/) driver used to connect. | `org.postgresql.Driver`; `com.microsoft.sqlserver.jdbc.SQLServerDriver` | Recommended |
+| `db.mysql.instance.address` | string | The hostname of the MySQL server instance bound to the current connection. This is useful in cases where the client is connecting to a MySQL server via a proxy or in a clustered environment. This client may obtain this value using queries like `select @@hostname`. | `mysql-e26b99z.example.com` | Recommended: If different from the `server.address` |
```
Do you, given the proposals in #321 (e. g. https://github.com/open-telemetry/semantic-conventions/pull/321#issuecomment-1720609696), that `server.address` could be sufficient to model this?

pyohannes replied:
Got it, so `db.mysql.instance.address` will identify the node inside the MySQL cluster. And `server.address` will be the address the user connects to.

arielvalentin replied:
Yes!

pyohannes commented on docs/database/mysql.md:
```diff
@@ -0,0 +1,24 @@
+
+**Status**: [Experimental][DocumentStatus]
+
+The Semantic Conventions for the [MySQL](https://www.mysql.com/) extend and override the [Database Semantic Conventions](database-spans.md)
```
```suggestion
The Semantic Conventions for the [MySQL](https://www.mysql.com/) extend the [Database Semantic Conventions](database-spans.md)
```

A nit, but I think product-specific semantic conventions shouldn't _override_ general domain-specific ones. General domain-specific semantic conventions should apply to all related products in the domain and allow consumers to interpret data without considering any product-specific details.

arielvalentin commented on docs/database/mysql.md:
```diff
@@ -0,0 +1,24 @@
+| Attribute  | Type | Description  | Examples  | Requirement Level |
+|---|---|---|---|---|
+| [`db.jdbc.driver_classname`](database-spans.md) | string | The fully-qualified class name of the [Java Database Connectivity (JDBC)](https://docs.oracle.com/javase/8/docs/technotes/guides/jdbc/) driver used to connect. | `org.postgresql.Driver`; `com.microsoft.sqlserver.jdbc.SQLServerDriver` | Recommended |
+| `db.mysql.instance.address` | string | The hostname of the MySQL server instance bound to the current connection. This is useful in cases where the client is connecting to a MySQL server via a proxy or in a clustered environment. This client may obtain this value using queries like `select @@hostname`. | `mysql-e26b99z.example.com` | Recommended: If different from the `server.address` |
```
Some other options I can think of for a more general attribute name:

- `db.instance.server.address`
- `db.node.server.address`

arielvalentin replied:
Naming is notoriously difficult so what do you all think of one of the following:

- `db.instance.address`
- `db.instance.server.address`
- `db.node.address`
- `db.node.server.address`
- `db.server.address`

pyohannes replied:
Mind that I'm not much of a database expert, but `db.instance.address` makes the most sense to me.

jsuereth replied:
I'd vote for `db.instance.address`

AlexanderWert replied:
@jsuereth Please see my comment here: https://github.com/open-telemetry/semantic-conventions/pull/345#discussion_r1393789543

arielvalentin commented:
@pyohannes @AlexanderWert Is there something more that you need from me? Are you wanting to explore the route of generic attribute naming?

Oberon00 commented:
As I understand the new address conventions

> The hostname of the MySQL server instance bound to the current connection. This is useful in cases where the client is connecting to a MySQL server via a proxy or in a clustered environment

This should be filled in in server.address. The current definition of server.address might need to be / have already implicitly changed. See #342.

CC @trask @lmolkova It seems to me that the new address conventions should make this PR redundant, maybe this request could inform / validate the design of #342.

pyohannes commented:
@Oberon00 

> This should be filled in in server.address. The current definition of server.address might need to be / have already implicitly changed. See https://github.com/open-telemetry/semantic-conventions/pull/342.

I had the same thought, but there's more to it: https://github.com/open-telemetry/semantic-conventions/pull/345#discussion_r1337154428:

> Got it, so `db.mysql.instance.address` will identify the node inside the MySQL cluster. And `server.address` will be the address the user connects to.

arielvalentin commented:
cc: @cschleiden @andyedison @adrianna-chang-shopify @open-telemetry/ruby-contrib-approvers

AlexanderWert commented on model/trace/database.yaml:
```diff
@@ -249,6 +249,16 @@ groups:
       - ref: server.socket.domain
         requirement_level:
           recommended: If different than `server.address` and if `server.socket.address` is set.
+      - id: instance.address
```
If we want to use that attribute in a generic way to identify an instance within a DB cluster, IMHO `db.instance.address` is not the best name, as it's too prescriptive of requiring an address.

For example we cannot use it to replace `db.mssql.instance_name` or `db.elasticsearch.node.name` as those are not necessarily adresses.

How about something more generic that would allow to put either an address, and id or name?

Something like `db.instance.id` ?

arielvalentin replied:
`db.instance.id` works for me.

AlexanderWert reviewed (changes requested):
See my inline comment regarding the name of the generic attribute.

Please, also resolve the conflicts and update this PR as some of the previous fields have changed (e.g. `server.socket.address`).

arielvalentin commented:
@AlexanderWert I have rebased and resolved conflicts as well as renamed the attribute to `db.instance.id`.

AlexanderWert commented on docs/database/database-spans.md:
```diff
@@ -65,6 +65,7 @@ Some database systems may allow a connection to switch to a different `db.user`,
 | Attribute  | Type | Description  | Examples  | Requirement Level |
 |---|---|---|---|---|
 | `db.connection_string` | string | The connection string used to connect to the database. It is recommended to remove embedded credentials. | `Server=(localdb)\v11.0;Integrated Security=true;` | Recommended |
+| `db.instance.id` | string | The db instance id that is executing queries or mutations on the current connection. This is useful in cases where the database is running in a clustered environment and the instrumentation is able to record the node executing the query. The client may obtain this value in databases like MySQL using queries like `select @@hostname`. | `mysql-e26b99z.example.com` | Re ...
```
```suggestion
| `db.instance.id` | string | An identifier (address, unique name, or any other identifier) of the database instance that is executing queries or mutations on the current connection. This is useful in cases where the database is running in a clustered environment and the instrumentation is able to record the node executing the query. The client may obtain this value in databases like MySQL using queries like `select @@hostname`. | `mysql-e26b99z.example.com` | Recommended: If different from the `server.address` |
```

AlexanderWert commented on model/trace/database.yaml:
```diff
@@ -246,6 +246,16 @@ groups:
+        tag: connection-level
+        type: string
+        brief: >
+          The db instance id that is executing queries or mutations on the current connection.
```
```suggestion
          An identifier (address, unique name, or any other identifier) of the database instance that is executing queries or mutations on the current connection.
```

AlexanderWert reviewed (approved):
Nit: a small suggestion on the wording to make clear this field can contain any type of identifier (address, unique name, ids, etc.)

arielvalentin commented:
@AlexanderWert thanks again for the review I have rebased and fixed the latest conflicts introduced by the registry refactoring.

Is there anything else that I need to do?

AlexanderWert commented:
> @AlexanderWert thanks again for the review I have rebased and fixed the latest conflicts introduced by the registry refactoring.
> 
> Is there anything else that I need to do?

LGTM!

We just need more approvals.
